[mono] Crash in roslyn due to null AssemblyLoadContext.GetLoadContext result

<!--This is just a template - feel free to delete any and all of it and replace as appropriate.-->

### Description

After upgrading my Mono-based runtime for s390x to preview6, using analyzers with roslyn now crashes, while it used to work in preview5. 

### Configuration

.NET 6 preview6 re-built to target s390x.

<!--
* Which version of .NET is the code running on?
* What OS and version, and what distro if applicable?
* What is the architecture (x64, x86, ARM, ARM64)?
* Do you know whether it is specific to that configuration?
* If you're using Blazor, which web browser(s) do you see this issue in?
  -->

### Regression?

Yes, this did work with preview5.   Reverting the patch from https://github.com/dotnet/runtime/pull/53308 fixes the problem.

<!--
* Did this work in a previous build or release of .NET Core, or from .NET Framework? If you can try a previous release or build to find out, that can help us narrow down the problem. If you don't know, that's OK.
  -->

### Other information

The symptom is error messages along those lines:

```
error CS8034: Unable to load Analyzer assembly /home/uweigand/.nuget/packages/microsoft.codeanalysis.csharp.codestyle/3.10.0/analyzers/dotnet/cs/Microsoft.CodeAnalysis.CSharp.CodeStyle.dll : Object reference not set to an instance of an object
error CS8034: Unable to load Analyzer assembly /home/uweigand/.nuget/packages/microsoft.codeanalysis.csharp.codestyle/3.10.0/analyzers/dotnet/cs/Microsoft.CodeAnalysis.CodeStyle.dll : Object reference not set to an instance of an object
error CS8034: Unable to load Analyzer assembly /home/uweigand/.nuget/packages/microsoft.codeanalysis.netanalyzers/6.0.0-rc1.21356.1/analyzers/dotnet/cs/Microsoft.CodeAnalysis.CSharp.NetAnalyzers.dll : Object reference not set to an instance of an object
error CS8034: Unable to load Analyzer assembly /home/uweigand/.nuget/packages/microsoft.codeanalysis.netanalyzers/6.0.0-rc1.21356.1/analyzers/dotnet/cs/Microsoft.CodeAnalysis.NetAnalyzers.dll : Object reference not set to an instance of an object
error CS8034: Unable to load Analyzer assembly /home/uweigand/.nuget/packages/microsoft.dotnet.codeanalysis/6.0.0-beta.21359.3/analyzers/Microsoft.DotNet.CodeAnalysis.dll : Object reference not set to an instance of an object
error CS8034: Unable to load Analyzer assembly /home/uweigand/.nuget/packages/microsoft.net.illink.analyzers/6.0.100-preview.6.21362.3/analyzers/dotnet/cs/ILLink.RoslynAnalyzer.dll : Object reference not set to an instance of an object
error CS8034: Unable to load Analyzer assembly /home/uweigand/.nuget/packages/stylecop.analyzers.unstable/1.2.0.304/analyzers/dotnet/cs/StyleCop.Analyzers.dll : Object reference not set to an instance of an object
```

triggered in roslyn's `Microsoft.CodeAnalysis.DefaultAnalyzerAssemblyLoader:LoadFromPathImpl` because `AssemblyLoadContext.GetLoadContext` now returns null for the default ALC.

This is related to https://github.com/dotnet/runtime/pull/53308.  Before that patch, the default ALC used to be initialized in `System.Runtime.Loader.AssemblyLoadContext:InternalInitializeNativeALC` when loading the `libSystem.Native` library on startup.

This routine was called out of the default `AssemblyLoadContext` constructor (which actually allocates the handle) invoked when `GetAssemblyLoadContext` is called with a `IntPtr.Zero` handler, which happens in the `MonoResolveUnmanagedDll` handler, which in turn was called from `netcore_resolve_with_load`.

However, the above PR added this early exit to `netcore_resolve_with_load`:

```
        if (!mono_gchandle_get_target_internal (alc->gchandle))
                return NULL;
```
and now `MonoResolveUnmanagedDll` is never called and therefore we never actually seem to get a non-null handle.

Not sure what the proper fix is, but reverting the PR makes the problem go away.

@CoffeeFlux, @vargaz, @lambdageek can you have a look?

<!--
* Please include any relevant stack traces or error messages. If possible please include text as text rather than images (so it shows up in searches).
* If you have an idea where the problem might lie, let us know that here. Please include any pointers to code, relevant changes, or related issues you know of.
* Do you know of any workarounds?
  -->

--- a/src/mono/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.Mono.cs
+++ b/src/mono/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.Mono.cs
@@ -126,24 +126,14 @@ namespace System.Runtime.Loader
         // success.
         private static Assembly? MonoResolveUsingResolvingEvent(IntPtr gchALC, string assemblyName)
         {
-            AssemblyLoadContext context;
-            // This check exists because the function can be called early in startup, before the default ALC is initialized
-            if (gchALC == IntPtr.Zero)
-                context = Default;
-            else
-                context = (AssemblyLoadContext)(GCHandle.FromIntPtr(gchALC).Target)!;
+            AssemblyLoadContext context = GetAssemblyLoadContext(gchALC);
             return context.ResolveUsingEvent(new AssemblyName(assemblyName));
         }
 
         // Invoked by Mono to resolve requests to load satellite assemblies.
         private static Assembly? MonoResolveUsingResolveSatelliteAssembly(IntPtr gchALC, string assemblyName)
         {
-            AssemblyLoadContext context;
-            // This check exists because the function can be called early in startup, before the default ALC is initialized
-            if (gchALC == IntPtr.Zero)
-                context = Default;
-            else
-                context = (AssemblyLoadContext)(GCHandle.FromIntPtr(gchALC).Target)!;
+            AssemblyLoadContext context = GetAssemblyLoadContext(gchALC);
             return context.ResolveSatelliteAssembly(new AssemblyName(assemblyName));
         }
 

--- a/src/mono/mono/metadata/assembly-load-context.c
+++ b/src/mono/mono/metadata/assembly-load-context.c
@@ -445,9 +445,6 @@ invoke_resolve_method (MonoMethod *resolve_method, MonoAssemblyLoadContext *alc,
 	if (mono_runtime_get_no_exec ())
 		return NULL;
 
-	if (!mono_gchandle_get_target_internal (alc->gchandle))
-		return NULL;
-
 	HANDLE_FUNCTION_ENTER ();
 
 	aname_str = mono_stringify_assembly_name (aname);
@@ -457,7 +454,14 @@ invoke_resolve_method (MonoMethod *resolve_method, MonoAssemblyLoadContext *alc,
 
 	MonoReflectionAssemblyHandle assm;
 	gpointer gchandle;
-	gchandle = (gpointer)alc->gchandle;
+	/* for the default ALC, pass NULL to ask for the Default ALC - see
+	 * AssemblyLoadContext.GetAssemblyLoadContext(IntPtr gchManagedAssemblyLoadContext) - which
+	 * will create the managed ALC object if it hasn't been created yet
+	 */
+	if (alc->gchandle == default_alc->gchandle)
+		gchandle = NULL;
+	else
+		gchandle = GUINT_TO_POINTER (alc->gchandle);
 	gpointer args [2];
 	args [0] = &gchandle;
 	args [1] = MONO_HANDLE_RAW (aname_obj);

--- a/src/mono/mono/metadata/assembly-load-context.c
+++ b/src/mono/mono/metadata/assembly-load-context.c
@@ -453,15 +453,7 @@ invoke_resolve_method (MonoMethod *resolve_method, MonoAssemblyLoadContext *alc,
 	goto_if_nok (error, leave);
 
 	MonoReflectionAssemblyHandle assm;
-	gpointer gchandle;
-	/* for the default ALC, pass NULL to ask for the Default ALC - see
-	 * AssemblyLoadContext.GetAssemblyLoadContext(IntPtr gchManagedAssemblyLoadContext) - which
-	 * will create the managed ALC object if it hasn't been created yet
-	 */
-	if (alc->gchandle == default_alc->gchandle)
-		gchandle = NULL;
-	else
-		gchandle = GUINT_TO_POINTER (alc->gchandle);
+	gpointer gchandle = mono_alc_get_gchandle_for_resolving (alc);
 	gpointer args [2];
 	args [0] = &gchandle;
 	args [1] = MONO_HANDLE_RAW (aname_obj);

--- a/src/mono/mono/metadata/loader-internals.h
+++ b/src/mono/mono/metadata/loader-internals.h
@@ -269,6 +269,19 @@ mono_alc_get_ambient (void)
 	return mono_alc_get_default ();
 }
 
+static inline MonoGCHandle
+mono_alc_get_gchandle_for_resolving (MonoAssemblyLoadContext *alc)
+{
+	/* for the default ALC, pass NULL to ask for the Default ALC - see
+	 * AssemblyLoadContext.GetAssemblyLoadContext(IntPtr gchManagedAssemblyLoadContext) - which
+	 * will create the managed ALC object if it hasn't been created yet
+	 */
+	if (alc->gchandle == mono_alc_get_default ()->gchandle)
+		return NULL;
+	else
+		return GUINT_TO_POINTER (alc->gchandle);
+}
+
 MonoAssemblyLoadContext *
 mono_alc_from_gchandle (MonoGCHandle alc_gchandle);
 

--- a/src/mono/mono/metadata/native-library.c
+++ b/src/mono/mono/metadata/native-library.c
@@ -698,9 +698,6 @@ netcore_resolve_with_resolving_event (MonoAssemblyLoadContext *alc, MonoAssembly
 	if (mono_runtime_get_no_exec ())
 		return NULL;
 
-	if (!mono_gchandle_get_target_internal (alc->gchandle))
-		return NULL;
-
 	HANDLE_FUNCTION_ENTER ();
 
 	MonoStringHandle scope_handle;

--- a/src/mono/mono/metadata/native-library.c
+++ b/src/mono/mono/metadata/native-library.c
@@ -628,9 +628,6 @@ netcore_resolve_with_load (MonoAssemblyLoadContext *alc, const char *scope, Mono
 	if (mono_runtime_get_no_exec ())
 		return NULL;
 
-	if (!mono_gchandle_get_target_internal (alc->gchandle))
-		return NULL;
-
 	HANDLE_FUNCTION_ENTER ();
 
 	MonoStringHandle scope_handle;
@@ -638,7 +635,14 @@ netcore_resolve_with_load (MonoAssemblyLoadContext *alc, const char *scope, Mono
 	goto_if_nok (error, leave);
 
 	gpointer gchandle;
-	gchandle = GUINT_TO_POINTER (alc->gchandle);
+	/* for the default ALC, pass NULL to ask for the Default ALC - see
+	 * AssemblyLoadContext.GetAssemblyLoadContext(IntPtr gchManagedAssemblyLoadContext) - which
+	 * will create the managed ALC object if it hasn't been created yet
+	 */
+	if (alc->gchandle == mono_alc_get_default ()->gchandle)
+		gchandle = NULL;
+	else
+		gchandle = GUINT_TO_POINTER (alc->gchandle);
 	gpointer args [3];
 	args [0] = MONO_HANDLE_RAW (scope_handle);
 	args [1] = &gchandle;
@@ -708,7 +712,14 @@ netcore_resolve_with_resolving_event (MonoAssemblyLoadContext *alc, MonoAssembly
 	goto_if_nok (error, leave);
 
 	gpointer gchandle;
-	gchandle = GUINT_TO_POINTER (alc->gchandle);
+	/* for the default ALC, pass NULL to ask for the Default ALC - see
+	 * AssemblyLoadContext.GetAssemblyLoadContext(IntPtr gchManagedAssemblyLoadContext) - which
+	 * will create the managed ALC object if it hasn't been created yet
+	 */
+	if (alc->gchandle == mono_alc_get_default ()->gchandle)
+		gchandle = NULL;
+	else
+		gchandle = GUINT_TO_POINTER (alc->gchandle);
 	gpointer args [4];
 	args [0] = MONO_HANDLE_RAW (scope_handle);
 	args [1] = MONO_HANDLE_RAW (assembly_handle);

--- a/src/mono/mono/metadata/native-library.c
+++ b/src/mono/mono/metadata/native-library.c
@@ -634,15 +634,7 @@ netcore_resolve_with_load (MonoAssemblyLoadContext *alc, const char *scope, Mono
 	scope_handle = mono_string_new_handle (scope, error);
 	goto_if_nok (error, leave);
 
-	gpointer gchandle;
-	/* for the default ALC, pass NULL to ask for the Default ALC - see
-	 * AssemblyLoadContext.GetAssemblyLoadContext(IntPtr gchManagedAssemblyLoadContext) - which
-	 * will create the managed ALC object if it hasn't been created yet
-	 */
-	if (alc->gchandle == mono_alc_get_default ()->gchandle)
-		gchandle = NULL;
-	else
-		gchandle = GUINT_TO_POINTER (alc->gchandle);
+	gpointer gchandle = mono_alc_get_gchandle_for_resolving (alc);
 	gpointer args [3];
 	args [0] = MONO_HANDLE_RAW (scope_handle);
 	args [1] = &gchandle;
@@ -708,15 +700,7 @@ netcore_resolve_with_resolving_event (MonoAssemblyLoadContext *alc, MonoAssembly
 	assembly_handle = mono_assembly_get_object_handle (assembly, error);
 	goto_if_nok (error, leave);
 
-	gpointer gchandle;
-	/* for the default ALC, pass NULL to ask for the Default ALC - see
-	 * AssemblyLoadContext.GetAssemblyLoadContext(IntPtr gchManagedAssemblyLoadContext) - which
-	 * will create the managed ALC object if it hasn't been created yet
-	 */
-	if (alc->gchandle == mono_alc_get_default ()->gchandle)
-		gchandle = NULL;
-	else
-		gchandle = GUINT_TO_POINTER (alc->gchandle);
+	gpointer gchandle = mono_alc_get_gchandle_for_resolving (alc);
 	gpointer args [4];
 	args [0] = MONO_HANDLE_RAW (scope_handle);
 	args [1] = MONO_HANDLE_RAW (assembly_handle);

